File: MacClient/Sources/CCAnywhere/Models/ProtocolMessage.swift
// ProtocolMessage.swift
// WebSocket protocol message envelopes (see 需求规格说明书 §3.4).
// We model the minimum set the MacClient needs to consume/emit.

import Foundation

/// All messages share this envelope.
public struct ProtocolMessage: Codable, Sendable {
    public let type: String
    public let id: String
    public let data: AnyJSON?

    public init(type: String, id: String = UUID().uuidString, data: AnyJSON? = nil) {
        self.type = type
        self.id = id
        self.data = data
    }
}

// MARK: - Outbound payloads

public struct BindMacRequest: Codable, Sendable {
    public let type: String     // "mac"
    public let token: String

    public init(token: String) {
        self.type = "mac"
        self.token = token
    }
}

public struct TabListPayload: Codable, Sendable {
    public let tabs: [TabSummary]
    public init(tabs: [TabSummary]) { self.tabs = tabs }
}

public struct MsgHistoryRequestPayload: Codable, Sendable {
    public let tabId: String
    public let limit: Int?
    public let before: String?

    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case limit
        case before
    }
}

public struct MsgHistoryResponsePayload: Codable, Sendable {
    public let tabId: String
    public let messages: [AnyJSON]
    public let hasMore: Bool

    public init(tabId: String, messages: [AnyJSON], hasMore: Bool) {
        self.tabId = tabId
        self.messages = messages
        self.hasMore = hasMore
    }

    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case messages
        case hasMore = "has_more"
    }
}

public struct TabSummary: Codable, Sendable {
    public let id: String
    public let name: String
    public let folder: String
    public let claudeStatus: String
    public let lastActivityAt: String?

    public init(id: String,
                name: String,
                folder: String,
                claudeStatus: String,
                lastActivityAt: String?) {
        self.id = id
        self.name = name
        self.folder = folder
        self.claudeStatus = claudeStatus
        self.lastActivityAt = lastActivityAt
    }

    private enum CodingKeys: String, CodingKey {
        case id, name, folder
        case claudeStatus = "claude_status"
        case lastActivityAt = "last_activity_at"
    }
}

public struct TabChangedPayload: Codable, Sendable {
    public let tab: TabSummary
    public let action: String   // "added" | "removed" | "renamed"

    public init(tab: TabSummary, action: String) {
        self.tab = tab
        self.action = action
    }
}

public struct MsgStreamPayload: Codable, Sendable {
    public let tabId: String
    public let messages: [AnyJSON]

    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case messages
    }
}

public struct DeviceRevokeRequest: Codable, Sendable {
    public let subTokenId: String
    private enum CodingKeys: String, CodingKey {
        case subTokenId = "sub_token_id"
    }
}

// MARK: - Inbound payloads (subset)

public struct InputTextPayload: Codable, Sendable {
    public let tabId: String
    public let text: String
    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case text
    }
}

public struct InputImagePayload: Codable, Sendable {
    public let tabId: String
    public let imageUrl: String
    public let filename: String
    public let sha256: String?
    public let uploadId: String
    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case imageUrl = "image_url"
        case filename
        case sha256
        case uploadId = "upload_id"
    }
}

public struct ToolUseApprovePayload: Codable, Sendable {
    public let tabId: String
    public let action: String   // "approve" | "reject" | "always_approve"
    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case action
    }
}

public struct BindAckPayload: Codable, Sendable {
    public let agentId: String?
    public let sessionToken: String?
    private enum CodingKeys: String, CodingKey {
        case agentId = "agent_id"
        case sessionToken = "session_token"
    }
}

public struct PhoneCountPayload: Codable, Sendable {
    public let count: Int
    public let names: [String]?
}

public struct DeviceBoundPayload: Codable, Sendable {
    public let subTokenId: String
    public let deviceName: String
    public let deviceModel: String?
    public let osVersion: String?
    public let boundAt: String?
    private enum CodingKeys: String, CodingKey {
        case subTokenId = "sub_token_id"
        case deviceName = "device_name"
        case deviceModel = "device_model"
        case osVersion = "os_version"
        case boundAt = "bound_at"
    }
}

public struct ErrorPayload: Codable, Sendable {
    public let code: String
    public let message: String
}

// MARK: - AskUserQuestion 远程交互 payloads

public struct AskQuestionOption: Codable, Sendable {
    public let label: String
    public let description: String?

    public init(label: String, description: String? = nil) {
        self.label = label
        self.description = description
    }
}

public struct AskQuestionItem: Codable, Sendable {
    public let question: String
    public let header: String
    public let multiSelect: Bool
    public let options: [AskQuestionOption]

    public init(question: String,
                header: String,
                multiSelect: Bool,
                options: [AskQuestionOption]) {
        self.question = question
        self.header = header
        self.multiSelect = multiSelect
        self.options = options
    }

    private enum CodingKeys: String, CodingKey {
        // multiSelect 用 camelCase（保持与 AskUserQuestion 工具原生 schema 一致，
        // 也与三端协议（需求规格说明书 §3.2.1）一致；该字段是 questions 数组元素
        // 内部字段，由 Claude SDK 写入 tool_input，必须保留 camelCase。
        case question
        case header
        case multiSelect
        case options
    }
}

public struct AskQuestionPendingPayload: Codable, Sendable {
    public let requestId: String
    public let tabId: String
    public let toolUseId: String
    public let askKind: String      // "user_question" | "tool_approval"
    public let allowOther: Bool
    public let questions: [AskQuestionItem]?
    public let toolName: String?
    public let toolInput: AnyJSON?

    public init(requestId: String,
                tabId: String,
                toolUseId: String,
                askKind: String,
                allowOther: Bool,
                questions: [AskQuestionItem]? = nil,
                toolName: String? = nil,
                toolInput: AnyJSON? = nil) {
        self.requestId = requestId
        self.tabId = tabId
        self.toolUseId = toolUseId
        self.askKind = askKind
        self.allowOther = allowOther
        self.questions = questions
        self.toolName = toolName
        self.toolInput = toolInput
    }

    private enum CodingKeys: String, CodingKey {
        case requestId = "request_id"
        case tabId = "tab_id"
        case toolUseId = "tool_use_id"
        case askKind = "ask_kind"
        case allowOther = "allow_other"
        case questions
        case toolName = "tool_name"
        case toolInput = "tool_input"
    }
}

public struct AskQuestionAnswerPayload: Codable, Sendable {
    public let requestId: String
    public let answers: [String: String]

    public init(requestId: String, answers: [String: String]) {
        self.requestId = requestId
        self.answers = answers
    }

    private enum CodingKeys: String, CodingKey {
        case requestId = "request_id"
        case answers
    }
}

/// Phone → Mac，tool_approval 决策回执（F4 危险工具远程批准）。
///
/// envelope.type = "ask.tool_approval.answer"。Mac 端
/// `DependencyContainer.handleAskToolApprovalInbound` 解析后投递给
/// `HookIpcServer.receiveApprovalFromWs`，进入 winner 锁仲裁。
/// 字段命名与 Server `protocol.AskToolApprovalAnswer` 严格对齐（snake_case）。
public struct AskToolApprovalAnswerPayload: Codable, Sendable {
    public let requestId: String
    /// "allow" | "deny"
    public let decision: String
    /// 用户附加的拒绝原因，可选（R-F4-005）。
    public let reason: String?

    public init(requestId: String, decision: String, reason: String?) {
        self.requestId = requestId
        self.decision = decision
        self.reason = reason
    }

    private enum CodingKeys: String, CodingKey {
        case requestId = "request_id"
        case decision
        case reason
    }
}

public struct AskQuestionAnsweredPayload: Codable, Sendable {
    public let requestId: String
    public let answeredBy: String
    public let answers: [String: String]

    public init(requestId: String, answeredBy: String, answers: [String: String]) {
        self.requestId = requestId
        self.answeredBy = answeredBy
        self.answers = answers
    }

    private enum CodingKeys: String, CodingKey {
        case requestId = "request_id"
        case answeredBy = "answered_by"
        case answers
    }
}

public struct AskQuestionTimeoutPayload: Codable, Sendable {
    public let requestId: String
    public let reason: String

    public init(requestId: String, reason: String) {
        self.requestId = requestId
        self.reason = reason
    }

    private enum CodingKeys: String, CodingKey {
        case requestId = "request_id"
        case reason
    }
}

public struct ToolProgressPrePayload: Codable, Sendable {
    public let tabId: String
    public let toolUseId: String
    public let toolName: String
    public let toolInput: AnyJSON

    public init(tabId: String,
                toolUseId: String,
                toolName: String,
                toolInput: AnyJSON) {
        self.tabId = tabId
        self.toolUseId = toolUseId
        self.toolName = toolName
        self.toolInput = toolInput
    }

    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case toolUseId = "tool_use_id"
        case toolName = "tool_name"
        case toolInput = "tool_input"
    }
}

public struct ToolProgressPostPayload: Codable, Sendable {
    public let tabId: String
    public let toolUseId: String
    public let toolName: String
    public let success: Bool
    public let error: String?

    public init(tabId: String,
                toolUseId: String,
                toolName: String,
                success: Bool,
                error: String? = nil) {
        self.tabId = tabId
        self.toolUseId = toolUseId
        self.toolName = toolName
        self.success = success
        self.error = error
    }

    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case toolUseId = "tool_use_id"
        case toolName = "tool_name"
        case success
        case error
    }
}

public struct NotificationPayload: Codable, Sendable {
    public let tabId: String
    public let notificationType: String
    public let title: String
    public let message: String

    public init(tabId: String,
                notificationType: String,
                title: String,
                message: String) {
        self.tabId = tabId
        self.notificationType = notificationType
        self.title = title
        self.message = message
    }

    private enum CodingKeys: String, CodingKey {
        case tabId = "tab_id"
        case notificationType = "notification_type"
        case title
        case message
    }
}

File: AndroidClient/lib/models/protocol_message.dart
/// WebSocket envelope
///
/// 协议见需求规格说明书 §3.4。统一形如：
/// `{"type": "...", "id": "uuid", "data": {...}}`
class ProtocolMessage {
  final String type;
  final String id;
  final Map<String, dynamic> data;

  const ProtocolMessage({
    required this.type,
    required this.id,
    this.data = const {},
  });

  Map<String, dynamic> toJson() => {
        'type': type,
        'id': id,
        'data': data,
      };

  static ProtocolMessage? tryFromJson(dynamic raw) {
    if (raw is! Map) return null;
    final m = raw.cast<String, dynamic>();
    final type = m['type'];
    final id = m['id'];
    if (type is! String || id is! String) return null;
    final data = m['data'];
    return ProtocolMessage(
      type: type,
      id: id,
      data: data is Map ? data.cast<String, dynamic>() : <String, dynamic>{},
    );
  }
}

/// 协议消息类型常量
abstract class ProtocolType {
  // 鉴权
  static const bind = 'bind';
  static const bindAck = 'bind.ack';
  static const bindError = 'bind.error';
  static const ping = 'ping';
  static const pong = 'pong';
  static const forceDisconnect = 'force_disconnect';

  // 设备
  static const deviceSelfUnbind = 'device.self_unbind';

  // Tab
  static const tabList = 'tab.list';
  static const tabListRequest = 'tab.list.request';
  static const tabListResponse = 'tab.list.response';
  static const tabChanged = 'tab.changed';
  static const tabActivity = 'tab.activity';

  // 消息
  static const msgStream = 'msg.stream';
  static const msgHistoryRequest = 'msg.history.request';
  static const msgHistoryResponse = 'msg.history.response';
  static const msgRaw = 'msg.raw';

  // 输入
  static const inputText = 'input.text';
  static const inputImage = 'input.image';
  static const imageFetched = 'image.fetched';
  static const inputError = 'input.error';
  static const imageUploadBegin = 'image.upload.begin';
  static const imageUploadUrl = 'image.upload.url';
  static const imageUploadExpired = 'image.upload.expired';
  static const imageDownloadRequest = 'image.download.url';
  static const imageDownloadResponse = 'image.download.url.response';
  // (no-op placeholder so #36 deps are independent of protocol changes)
  static const slashListRequest = 'slash.list.request';
  static const slashListResponse = 'slash.list.response';
  static const toolUseApprove = 'tool_use.approve';

  // presence
  static const presenceMacOnline = 'presence.mac_online';
  static const presenceMacOffline = 'presence.mac_offline';
  static const presencePhoneCount = 'presence.phone_count';

  // 4.7 Hook 实时桥接（AskUserQuestion 远程交互）
  static const askQuestionPending = 'ask.question.pending';
  static const askQuestionAnswer = 'ask.question.answer';
  static const askQuestionAnswered = 'ask.question.answered';
  static const askQuestionTimeout = 'ask.question.timeout';
  /// F4 危险工具远程批准回执：phone → server → mac
  static const askToolApprovalAnswer = 'ask.tool_approval.answer';
  static const toolProgressPre = 'tool.progress.pre';
  static const toolProgressPost = 'tool.progress.post';
  static const notification = 'notification';
}

/// `notification` payload —— Mac App 桥接 Claude Notification hook,
/// Server 广播给所有 phone（mac → server → phone）。
///
/// 字段命名严格对齐 Server `protocol.Notification` (snake_case JSON tag)。
/// 协议见需求规格说明书 §3.2.7、§3.4.4，业务规则 R-F3-001 ~ R-F3-003。
class NotificationPayload {
  /// tab UUID（用于解析展示 [Tab xxx] 前缀）
  final String tabId;

  /// 枚举：`idle` / `permission_prompt` / `error`
  /// 决定 toast 颜色（R-F3-003：灰 / 黄 / 红）。
  final String notificationType;

  /// 通知标题（如 "Claude idle"）
  final String title;

  /// 通知正文
  final String message;

  const NotificationPayload({
    required this.tabId,
    required this.notificationType,
    required this.title,
    required this.message,
  });

  static NotificationPayload? tryFrom(Map<String, dynamic> data) {
    final tabId = data['tab_id'];
    final type = data['notification_type'];
    if (tabId is! String || type is! String) return null;
    final title = data['title'];
    final message = data['message'];
    return NotificationPayload(
      tabId: tabId,
      notificationType: type,
      title: title is String ? title : '',
      message: message is String ? message : '',
    );
  }
}

/// `ask.question.pending` payload —— Mac App 发起、Server 广播给所有 phone。
///
/// 字段命名严格对齐 Server `protocol.AskQuestionPending` (snake_case JSON tag)。
/// 需求规格说明书 §3.2.1 / R-F1-012：`allow_other` 字段缺失时默认为 true，
/// phone 端必须始终展示"自定义回答"输入项。
class AskQuestionPendingPayload {
  final String requestId;
  final String tabId;
  final String toolUseId;
  /// "user_question" | "tool_approval"
  final String askKind;
  final bool allowOther;
  /// 形如 `[{ question, header, multiSelect, options: [{ label, description }] }, ...]`
  /// 与 message.dart 的 AskUserQuestion 事后模式 questions 字段同结构。
  final List<Map<String, dynamic>> questions;
  final String? toolName;
  final Map<String, dynamic>? toolInput;
  // ⊕ R-F5: 子 agent 上下文（旧版 mac 端不发，全部 null → UI 降级）
  final String? parentToolUseId;
  final String? subAgentSummary;
  final bool isFromSubAgent;

  const AskQuestionPendingPayload({
    required this.requestId,
    required this.tabId,
    required this.toolUseId,
    required this.askKind,
    required this.allowOther,
    required this.questions,
    this.toolName,
    this.toolInput,
    this.parentToolUseId,
    this.subAgentSummary,
    this.isFromSubAgent = false,
  });

  static AskQuestionPendingPayload? tryFrom(Map<String, dynamic> data) {
    final requestId = data['request_id'];
    if (requestId is! String || requestId.isEmpty) return null;
    final tabId = data['tab_id'] as String? ?? '';
    final toolUseId = data['tool_use_id'] as String? ?? '';
    final askKind = data['ask_kind'] as String? ?? 'user_question';
    // R-F1-012:字段缺失则默认 true(始终展示自定义回答输入项)
    final allowOther = data['allow_other'] as bool? ?? true;
    final rawQs = data['questions'];
    final qs = <Map<String, dynamic>>[];
    if (rawQs is List) {
      for (final q in rawQs) {
        if (q is Map) qs.add(q.cast<String, dynamic>());
      }
    }
    final toolInput = data['tool_input'];
    return AskQuestionPendingPayload(
      requestId: requestId,
      tabId: tabId,
      toolUseId: toolUseId,
      askKind: askKind,
      allowOther: allowOther,
      questions: qs,
      toolName: data['tool_name'] as String?,
      toolInput: toolInput is Map
          ? toolInput.cast<String, dynamic>()
          : null,
      parentToolUseId: data['parent_tool_use_id'] as String?,
      subAgentSummary: data['sub_agent_summary'] as String?,
      isFromSubAgent: (data['is_from_sub_agent'] as bool?) ?? false,
    );
  }
}

/// `ask.question.answer` payload —— phone 提交给 Mac App。
///
/// answers 字段:key 是 question 原文,value 是用户选择的 label
/// 字符串或自定义回答(Other)的任意字符串。R-F1-014 不区分类型。
class AskQuestionAnswerPayload {
  final String requestId;
  final Map<String, String> answers;

  const AskQuestionAnswerPayload({
    required this.requestId,
    required this.answers,
  });

  Map<String, dynamic> toJson() => {
        'request_id': requestId,
        'answers': answers,
      };
}

/// `ask.tool_approval.answer` payload —— phone 提交给 Mac App（F4 危险工具远程批准）。
///
/// 字段命名严格对齐 Server `protocol.AskToolApprovalAnswer` 与 Mac
/// `AskToolApprovalAnswerPayload`（snake_case）。Mac 端
/// `DependencyContainer.handleAskToolApprovalInbound` 解析后会调
/// `HookIpcServer.receiveApprovalFromWs(requestId, decision, reason)`。
///
/// decision 必填 "allow" | "deny"；reason 可选，仅在 deny 时填用户附加原因
/// （R-F4-005）。本类只承担"出站序列化"，不参与入站反序列化。
class AskToolApprovalAnswerPayload {
  final String requestId;
  /// "allow" | "deny"
  final String decision;
  /// 可选用户附加原因，仅 deny 路径有意义；空串/null 都不写入 JSON。
  final String? reason;

  const AskToolApprovalAnswerPayload({
    required this.requestId,
    required this.decision,
    this.reason,
  });

  Map<String, dynamic> toJson() => {
        'request_id': requestId,
        'decision': decision,
        if (reason != null && reason!.isNotEmpty) 'reason': reason,
      };
}

/// `tab.activity` payload —— Mac → server → phone，按 tab 推送 Claude 活动状态。
/// 由 Mac 端 hook 桥接驱动（PreToolUse → working / Notification idle → waiting），
/// 只在状态真发生变化时推送（增量）。
class TabActivityPayload {
  final String tabId;
  /// "working" | "waiting"
  final String activity;

  const TabActivityPayload({required this.tabId, required this.activity});

  static TabActivityPayload? tryFrom(Map<String, dynamic> data) {
    final tabId = data['tab_id'];
    if (tabId is! String || tabId.isEmpty) return null;
    final activity = (data['activity'] as String?) ?? 'waiting';
    return TabActivityPayload(tabId: tabId, activity: activity);
  }
}

/// `ask.question.answered` payload —— Mac App winner 仲裁后广播给所有 phone。
class AskQuestionAnsweredPayload {
  final String requestId;
  final String answeredBy;
  final Map<String, String> answers;

  const AskQuestionAnsweredPayload({
    required this.requestId,
    required this.answeredBy,
    required this.answers,
  });

  static AskQuestionAnsweredPayload? tryFrom(Map<String, dynamic> data) {
    final requestId = data['request_id'];
    if (requestId is! String || requestId.isEmpty) return null;
    final answeredBy = data['answered_by'] as String? ?? '';
    final rawAnswers = data['answers'];
    final answers = <String, String>{};
    if (rawAnswers is Map) {
      rawAnswers.forEach((k, v) {
        if (k is String && v is String) answers[k] = v;
      });
    }
    return AskQuestionAnsweredPayload(
      requestId: requestId,
      answeredBy: answeredBy,
      answers: answers,
    );
  }
}

/// `ask.question.timeout` payload —— Mac App 在 5 分钟内无人回答时广播,
/// 所有 phone 撤销卡片。
class AskQuestionTimeoutPayload {
  final String requestId;
  /// "timeout" | "cancelled"
  final String reason;

  const AskQuestionTimeoutPayload({
    required this.requestId,
    required this.reason,
  });

  static AskQuestionTimeoutPayload? tryFrom(Map<String, dynamic> data) {
    final requestId = data['request_id'];
    if (requestId is! String || requestId.isEmpty) return null;
    return AskQuestionTimeoutPayload(
      requestId: requestId,
      reason: data['reason'] as String? ?? 'timeout',
    );
  }
}

/// `tool.progress.pre` payload —— Mac App 在 PreToolUse hook 命中时广播给所有 phone。
///
/// 协议见需求规格说明书 §3.2.5，业务规则 R-F2-001 ~ R-F2-005。
/// `tool_input` 中的长字段（command / file_path / content）由 Mac 端截断至 200 字符
/// （R-F2-004），phone 端再按 tool_name 提取摘要并二次截断 60 字符展示。
class ToolProgressPrePayload {
  final String tabId;
  final String toolUseId;
  final String toolName;
  /// 原始 tool_input（已由 Mac 端截断长字段至 200 字符）。
  final Map<String, dynamic> toolInput;

  const ToolProgressPrePayload({
    required this.tabId,
    required this.toolUseId,
    required this.toolName,
    required this.toolInput,
  });

  static ToolProgressPrePayload? tryFrom(Map<String, dynamic> data) {
    final toolUseId = data['tool_use_id'];
    if (toolUseId is! String || toolUseId.isEmpty) return null;
    final toolName = data['tool_name'];
    if (toolName is! String || toolName.isEmpty) return null;
    final input = data['tool_input'];
    return ToolProgressPrePayload(
      tabId: data['tab_id'] as String? ?? '',
      toolUseId: toolUseId,
      toolName: toolName,
      toolInput: input is Map ? input.cast<String, dynamic>() : <String, dynamic>{},
    );
  }
}

/// `tool.progress.post` payload —— Mac App 在 PostToolUse hook 命中时广播。
///
/// 协议见需求规格说明书 §3.2.6。`success=true` → phone 端移除进度条；
/// `success=false` → 变红色 toast 显示 5 秒后消失（场景 F2-S3、§3.4.3）。
class ToolProgressPostPayload {
  final String tabId;
  final String toolUseId;
  final String toolName;
  final bool success;
  /// 失败时的错误描述；成功时为 null 或空串。
  final String? error;

  const ToolProgressPostPayload({
    required this.tabId,
    required this.toolUseId,
    required this.toolName,
    required this.success,
    this.error,
  });

  static ToolProgressPostPayload? tryFrom(Map<String, dynamic> data) {
    final toolUseId = data['tool_use_id'];
    if (toolUseId is! String || toolUseId.isEmpty) return null;
    final toolName = data['tool_name'] as String? ?? '';
    final success = data['success'] as bool? ?? true;
    final rawErr = data['error'];
    final err = rawErr is String && rawErr.isNotEmpty ? rawErr : null;
    return ToolProgressPostPayload(
      tabId: data['tab_id'] as String? ?? '',
      toolUseId: toolUseId,
      toolName: toolName,
      success: success,
      error: err,
    );
  }
}

/// 错误码常量
abstract class ProtocolErrorCode {
  static const invalidToken = 'INVALID_TOKEN';
  static const tokenExpired = 'TOKEN_EXPIRED';
  static const revoked = 'REVOKED';
  static const macOffline = 'MAC_OFFLINE';
  static const tabNotFound = 'TAB_NOT_FOUND';
  static const imageTooLarge = 'IMAGE_TOO_LARGE';
  static const sha256Mismatch = 'SHA256_MISMATCH';
  static const internal = 'INTERNAL';
}
